File: java/jaxrs/src/main/java/sample/web/HelloResource.java
package sample.web;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.QueryParam;


@Path("hello")
public class HelloResource {

    @QueryParam("hoge")
    private Hoge hoge;

    @QueryParam("fuga")
    private Fuga fuga;

    @GET
    public String hello() {
        return this.hoge + " : " + this.fuga;
    }

    public static class Hoge {

        private String value;

        public Hoge(String value) {
            System.out.println("Constructor");
            this.value = value;
        }

//        public static Hoge valueOf(String value) {
//            System.out.println("valueOf()");
//            return new Hoge(value);
//        }

        public static Hoge fromString(String value) {
            System.out.println("fromString");
            return new Hoge(value);
        }

        @Override
        public String toString() {
            return "Hoge(" + this.value + ")";
        }
    }
}

File: kotlin/javaee/src/main/kotlin/sample/web/HelloResource.kt
package sample.web

import sample.ejb.JavaEjb
import sample.ejb.KotlinEjb
import javax.ejb.EJB
import javax.enterprise.context.RequestScoped
import javax.inject.Inject
import javax.ws.rs.Path
import javax.ws.rs.GET
import javax.ws.rs.QueryParam


@Path("kotlin")
@RequestScoped
open class HelloResource {

    @Inject
    lateinit private var javaEjbAtInject: JavaEjb
    @EJB
    lateinit private var javaEjbAtEJB: JavaEjb
    @Inject
    lateinit private var kotlinEjbAtInject: KotlinEjb
    @EJB
    lateinit private var kotlinEjbAtEJB: KotlinEjb

    @GET
    open fun hello() : String {
        this.javaEjbAtInject.hello("javaEjbAtInject")
        this.javaEjbAtEJB.hello("javaEjbAtEJB")
        this.kotlinEjbAtInject.hello("kotlinEjbAtInject")
        this.kotlinEjbAtEJB.hello("kotlinEjbAtEJB")

        return "Hello Kotlin Java EE!!"
    }
}